### PR TITLE
feat(security): Update docker-compose security for Consul ACL Setup

### PR DESCRIPTION
**Pull request by jim-wang-intel**
- Add ENABLE_REGISTRY_ACL env
- tuned volumes

Signed-off-by: Jim Wang <yutsung.jim.wang@intel.com>

## PR Checklist
Please check if your PR fulfills the following requirements:

- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**If your build fails** due to your commit message not passing the build checks, please review the guidelines here: https://github.com/edgexfoundry/edgex-examples/blob/master/.github/CONTRIBUTING.md.

## What is the current behavior?

No Consul ACL enabled yet

## What is the new behavior?
Enable Consul ACL in docker

## Does this PR introduce a breaking change?

- [ ] Yes
- [x ] No

## New Imports
Are there any new imports or modules? If so, what are they used for and why?

## Specific Instructions
Are there any specific instructions or things that should be known prior to reviewing?

## Other information

### CODE REVIEW

**Review by lenny-intel — CHANGES_REQUESTED** [review]
Changes look good, but you need to run `make build`

**Comment by jim-wang-intel** [comment]
> Changes look good, but you need to run `make build`

yeap, will do once that edgex-go PR is merged.

**Review by lenny-intel — DISMISSED** [review]
LGTM

File: compose-builder/add-security.yml
Comment (bnevis-i) [inline compose-builder/add-security.yml:75]: Why do we need write access to these volumes now?

File: compose-builder/common-security.env
Comment (bnevis-i) [inline compose-builder/common-security.env:20]: I thought we were going to leave this as FALSE until we knew we could get it all in in this release?

File: compose-builder/add-security.yml
Comment (jim-wang-intel) [reply compose-builder/add-security.yml:75]: we save the bootstrap token into this volume

File: compose-builder/common-security.env
Comment (jim-wang-intel) [reply compose-builder/common-security.env:20]: the value to turn on of not is release time determination; right now we just want to go through the TAF test firsts.

File: compose-builder/add-security.yml
Comment (bnevis-i) [reply compose-builder/add-security.yml:75]: What about edgex-init?  What do we write there?

Also, I think it would be safer to write the bootstrap token to secrets/edge-kong than give edgex-kong access to read secrets/edgex-consul volume to read the bootstrap token for purposes of setting up the reverse proxy.

Comment (jim-wang-intel) [reply compose-builder/add-security.yml:75]: > What about edgex-init? What do we write there?

the sentinel file, edgex-init is regular volume, so we want to write there in the case of cleaning up docker's regular volume via `docker volume prune`. If we were writing it to edgex-consul, then it would not get clean up since that one is bind-mount volume to the host system, so for state-consistency  issue, we decide to write it to this regular docker volume.


> 
> Also, I think it would be safer to write the bootstrap token to secrets/edge-kong than give edgex-kong access to read secrets/edgex-consul volume to read the bootstrap token for purposes of setting up the reverse proxy. Consul is an "internal" service that is more trusted than an externally facing service.

not quite understand why it is safer in kong's volume, it has to be written volume anayway

Comment (bnevis-i) [reply compose-builder/add-security.yml:75]: > the sentinel file, edgex-init is regular volume, so we want to write there in the case of cleaning up docker's regular volume via `docker volume prune`. If we were writing it to edgex-consul, then it would not get clean up since that one is bind-mount volume to the host system, so for state-consistency issue, we decide to write it to this regular docker volume.

Why not write it to /consul/config or /consul/data which is already writable and holds consul-related data, rather than make a volume that is read-only writable to hold consul-specific data?

> not quite understand why it is safer in kong's volume, it has to be written volume anayway

Giving kong access to consul's secrets volume allows kong to impersonate consul.  Giving consul access to kong's volume allows consul to impersonate kong.  Nobody (except proxy-setup) depends on kong.  Everyone depends on consul.

Comment (jim-wang-intel) [reply compose-builder/add-security.yml:75]: i've sent you a meeting invite tomorrow to discuss bootstrapping token volume location. not sure i can comprehend all your answers

Comment (jim-wang-intel) [reply compose-builder/add-security.yml:75]: A note for a few factors to consider for our meeting discussion:
1. sentinel file should be in regular volume so that when consul's volume is pruned, it is gone, too
2. sentinel file by default is written to `/egex-init/consul-bootstrapper` in toml file, so we could potentially make only that volume to be writable instead of the whole /edgex-init
3. the consul process itself doesn't use edgex-consul secret volume at all
4. the secret volume is not a regular docker volume, it is bind-mount type of docker volume to the host system
5. right now, `edgex-consul` secret volume is only shared between services `secretstore-setup` and `consul`
6. `/tmp/` in linux system potentially gets cleaned up on every system reboot (depending on $TMPTIME, which usually is 0). So **will it be ok to allow  consul's bootstrap token be cleaned up and gone after reboot?** (note that it is the same principle applied to all edgex secrets tokens that are written into /tmp/ dir)
5. Q: do we need the bootstrap token later?  A: Maybe, GUI needs a token to authenticate once the default_policy in consul changed to "deny".  Whether using this token or not is a different issue.

Comment (jim-wang-intel) [reply compose-builder/add-security.yml:75]: - we can put sentinel file in `consul/config`
- we want to write this bootstrap consul token into a separate bind-mount volume specifically for kong to use
- to put this token on the new regular docker volume (volume to be shared with proxy-setup) so that it won't suffer the reboot consequence, but user may not be easy to get it
- make it easy to get it , possibly to expand the current secret-config utility to retrieve consul bootstrap token

Comment (jim-wang-intel) [reply compose-builder/add-security.yml]: implemented

**Review by lenny-intel — APPROVED** [review]
lgtm

**Review by bnevis-i — APPROVED** [review]

**Comment by lenny-intel** [comment]
This PR was merged into https://github.com/edgexfoundry/edgex-compose/pull/20

**Pull request closed without merging**
